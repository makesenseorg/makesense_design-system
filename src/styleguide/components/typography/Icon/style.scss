.ds-icon {
  @include reset;
  display: inline-flex;
  align-items: center;
  vertical-align: middle;
  height: 1em;
}

.ds-icon-svg {
  line-height: 1;
  height: 1.2em;
  // Use this if the icons are build with strokes
  // stroke: currentColor;
  // stroke-width: 2.5px;
  // stroke-linejoin: miter;
  // stroke-linecap: miter;
  // overflow: visible;
  // Use this if the icons are build with solids
  fill: currentColor
}