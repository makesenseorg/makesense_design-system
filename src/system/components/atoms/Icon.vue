<template>
  <font-awesome-icon
    v-if="library === 'font-awesome'"
    :icon="getFaIcon"
    :class="`icon icon--color-${color}`"
  />
  <svg
    v-else
    :class="`feather feather-${getIcon}`"
    xmlns="http://www.w3.org/2000/svg"
    :width="size"
    :height="size"
    :viewBox="`0 0 ${defaultSize} ${defaultSize}`"
    fill="none"
    stroke="currentColor"
    :stroke-width="strokeWidth"
    stroke-linecap="round"
    stroke-linejoin="round"
    v-html="path"
  ></svg>
</template>

<script>
// todo : import dynamically (without breaking use in nuxt.js)
import feather from "feather-icons";
import { FontAwesomeIcon } from "@fortawesome/vue-fontawesome";

import {
  faArrowDown,
  faSort,
  faBars,
  faSearch,
  faArrowUp,
  faTimes,
  faTag,
  faArrowRight,
  faArrowLeft,
  faLock,
  faLifeRing,
  faBuilding,
  faCheck,
  faGlobeEurope,
  faClock,
  faLink,
  faWallet,
  faShare,
  faMapMarkerAlt
} from "@fortawesome/free-solid-svg-icons";

/**
 * Nice icons to make it more fun âœ¨
 * @version 0.1.0
 */
export default {
  name: "MksIcon",
  components: { FontAwesomeIcon },
  props: {
    /**
     * Type of icon depending on the library used. Prefer feather when possible
     */
    type: {
      type: String,
      default: ""
    },
    /**
     * Color of the icon amongs theme colors, inherits from parent by default.
     */
    color: {
      type: String,
      default: "inherit"
    },
    /**
     * Which library to use : "font-awesome", "feather"
     */
    library: {
      type: String,
      default: "feather"
    },
    size: {
      type: String,
      default: "24"
    },
    strokeWidth: {
      type: String,
      default: "2"
    }
  },
  data() {
    return {
      defaultSize: 24,
      path: ""
    };
  },
  async created() {
    if (this.library === "feather") {
      this.path = feather.icons[this.type];
    }
  },
  computed: {
    getIcon: function() {
      if (this.type === "close") return "x";
      if (this.type === "arrowUp") return "arrow-up";
      if (this.type === "arrowDown") return "arrow-down";
      if (this.type === "arrowLeft") return "arrow-left";
      if (this.type === "arrowRight") return "arrow-right";
      return this.type;
    },

    getFaIcon: function() {
      switch (this.type) {
        case "arrowDown":
          return faArrowDown;
        case "arrowUp":
          return faArrowUp;
        case "arrowRight":
          return faArrowRight;
        case "arrowLeft":
          return faArrowLeft;
        case "close":
          return faTimes;
        case "search":
          return faSearch;
        case "share":
          return faShare;
        case "check":
          return faCheck;
        case "sort":
          return faSort;
        case "link":
          return faLink;
        case "company":
          return faBuilding;
        case "lang":
          return faGlobeEurope;
        case "clock":
          return faClock;
        case "mapMarker":
          return faMapMarkerAlt;
        case "tag":
          return faTag;
        case "wallet":
          return faWallet;
        case "lock":
          return faLock;
        case "menu":
          return faBars;
        case "help":
          return faLifeRing;
        default:
          return faSearch;
      }
    }
  }
};
</script>

<style lang="scss" scoped>
.icon {
  color: inherit;
  vertical-align: middle;
  font-size: inherit;
  width: 1em;

  &--color-neutral {
    color: $color-neutral-70;
  }

  &--color-primary {
    color: $color-primary;
  }

  &--color-secondary {
    color: $color-secondary;
  }

  &--color-tertiary {
    color: $color-tertiary;
  }

  &--color-positive {
    color: $color-success;
  }

  &--color-warning {
    color: $color-warning;
  }

  &--color-negative {
    color: $color-danger;
  }
}
</style>
<docs>
## Feather icons (default)
List of all available icons : https://feathericons.com/

Some examples :

```jsx
<p>
  <mks-icon type="activity"></mks-icon> 
  <mks-icon type="alert-triangle"></mks-icon> 
  <mks-icon type="arrow-left"></mks-icon>
  <mks-icon type="arrow-left-circle"></mks-icon>
  <mks-icon type="bell"></mks-icon> 
  <mks-icon type="check"></mks-icon> 
  <mks-icon type="clock"></mks-icon> 
  <mks-icon type="eye-off"></mks-icon> 
  <mks-icon type="filter"></mks-icon> 
  <mks-icon type="heart"></mks-icon> 
  <mks-icon type="list"></mks-icon> 
  <mks-icon type="lock"></mks-icon>
  <mks-icon type="map-pin"></mks-icon>
  <mks-icon type="menu"></mks-icon>
  <mks-icon type="save"></mks-icon>
  <mks-icon type="star"></mks-icon> 
  <mks-icon type="thumbs-up"></mks-icon> 
  <mks-icon type="user"></mks-icon>
  <mks-icon type="zoom-in"></mks-icon>
</p>
```

## Font Awesome icons 
Font awesome icons are still available. However you have to specify to use the font-awesome library. 

Only these ones are available as of now : 

```jsx
<p>
  <mks-icon type="arrowUp" library="font-awesome"></mks-icon> 
  <mks-icon type="arrowDown" library="font-awesome"></mks-icon> 
  <mks-icon type="arrowLeft" library="font-awesome"></mks-icon>
  <mks-icon type="arrowRight" library="font-awesome"></mks-icon>
  <mks-icon type="close" library="font-awesome"></mks-icon> 
  <mks-icon type="mapMarker" library="font-awesome"></mks-icon> 
  <mks-icon type="lang" library="font-awesome"></mks-icon> 
  <mks-icon type="link" library="font-awesome"></mks-icon> 
  <mks-icon type="tag" library="font-awesome"></mks-icon> 
  <mks-icon type="company" library="font-awesome"></mks-icon> 
  <mks-icon type="search" library="font-awesome"></mks-icon> 
  <mks-icon type="check" library="font-awesome"></mks-icon>
  <mks-icon type="share" library="font-awesome"></mks-icon>
  <mks-icon type="sort"  library="font-awesome"></mks-icon>
  <mks-icon type="clock" library="font-awesome"></mks-icon>
  <mks-icon type="wallet" library="font-awesome"></mks-icon> 
  <mks-icon type="lock" library="font-awesome"></mks-icon> 
  <mks-icon type="menu" library="font-awesome"></mks-icon>
  <mks-icon type="help" library="font-awesome"></mks-icon>
</p>
```

## Colors 

Inherits from parent by default 

```jsx
<p>
  <mks-icon type="arrow-up" color="neutral"></mks-icon> 
  <mks-icon type="arrow-up" color="primary"></mks-icon> 
  <mks-icon type="arrow-up" color="secondary"></mks-icon> 
  <mks-icon type="arrow-up" color="tertiary"></mks-icon> 
  <mks-icon type="arrow-up" color="positive"></mks-icon> 
  <mks-icon type="arrow-up" color="warning"></mks-icon> 
  <mks-icon type="arrow-up" color="negative"></mks-icon> 
</p>
```
</docs>
