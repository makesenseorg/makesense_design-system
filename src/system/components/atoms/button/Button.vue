<template>
  <component
    :is="tag"
    :class="[
      `button button--type-${type} button--size-${size}`,
      { 'button--loading': loading }
    ]"
    :role="role"
    tabindex="0"
    :type="inputType"
    :disabled="disabled"
    ref="container"
    :style="{
      minWidth: $slots.loading === undefined ? minWidth + 'px' : 'auto'
    }"
    @click="onClick"
    @keyup.enter="onClick"
  >
    <mks-icon
      v-if="icon && !iconRight"
      :type="icon"
      class="button__icon button__icon--left"
      :size="size === 'round' ? 24 : 20"
    ></mks-icon>
    <!-- @slot Content of the button-->
    <slot v-if="!loading" />
    <!-- @slot Content of the button when `loading` prop is true -->
    <slot v-else name="loading">
      <mks-loading
        type="standalone"
        :inline="true"
        color="#fff"
        :size="20"
      ></mks-loading>
    </slot>
    <mks-icon
      v-if="icon && iconRight"
      :type="icon"
      class="button__icon button__icon--right"
      :size="size === 'round' ? 24 : 20"
    ></mks-icon>
  </component>
</template>
<script>
/**
 * Buttons can be used for any action or inner link.
 * @version 0.3.0
 */
export default {
  name: "MksButton",
  release: "0.3.0",
  props: {
    /**
     * Buttons can be replaced by any of these tags : "div", "a", "input", "button", "span"
     */
    tag: {
      type: String,
      default: "button",
      validator: function(value) {
        return ["div", "a", "input", "button", "span"].indexOf(value) !== -1;
      }
    },
    /**
     * If loading is set to true, the button will display the loading slot or "Processing..."
     */
    loading: {
      type: Boolean,
      default: false
    },
    /**
     * If disabled is set to true, the button will be disabled
     */
    disabled: {
      type: Boolean,
      default: false
    },
    /**
     * Can set the input type, in case of button tag (ex: "submit")
     */
    inputType: {
      type: String,
      required: false,
      default: ""
    },
    /**
     * Color of the button. "primary",
            "secondary",
            "tertiary",
            "neutral",
            "positive",
            "warning",
            "negative"
     */
    type: {
      type: String,
      default: "primary",
      validator: function(value) {
        return (
          [
            "primary",
            "secondary",
            "tertiary",
            "neutral",
            "positive",
            "warning",
            "negative",
            "text",
            "text-active"
          ].indexOf(value) !== -1
        );
      }
    },
    /**
     * Has to be a valid icon type
     */
    icon: {
      type: String
    },
    /**
     * Place the icon on the right side of the text
     */
    iconRight: {
      type: Boolean,
      default: false
    },
    /**
     * Size of the button. "default", "small", "medium" "full", "round"
     */
    size: {
      type: String,
      default: "default",
      validator: function(value) {
        return (
          ["default", "small", "medium", "full", "round", "square"].indexOf(
            value
          ) !== -1
        );
      }
    },
    /**
     * Size of the button. "default", "small", "full", "round"
     */
    role: {
      type: String,
      default: "button"
    }
  },
  data: function() {
    return {
      minWidth: 0,
      observer: null
    };
  },
  mounted() {
    this.updateMinWidth();

    this.observer = new MutationObserver(() => {
      this.updateMinWidth();
    });

    this.observer.observe(this.$refs.container, {
      attributes: true,
      childList: true,
      characterData: true,
      subtree: true
    });
  },
  beforeDestroy: function() {
    // Clean up
    this.observer.disconnect();
  },
  methods: {
    onClick() {
      /**
       * Fires after click.
       *
       * @event click
       */
      this.$emit("click");
    },
    updateMinWidth() {
      this.$nextTick(function() {
        var container = this.$refs.container;
        if (container !== undefined) this.minWidth = container.offsetWidth;
      });
    }
  }
};
</script>

<style lang="scss" scoped>
.button {
  @include cursor-pointer;
  @include border-radius;
  @include button-reset;
  @include padding-button;
  @include text-body;
  display: inline-block;
  font-weight: $font-weight-extrabold; // override text-body
  border-width: $border-width-m;
  border-style: solid;
  text-decoration: none !important;
  transition: background-color 0.2s, border-color 0.2s;
  text-align: center;

  &:focus {
    @include focus($background-color-primary);
  }

  &--size-small {
    @include border-round;
    @include padding-button-small;
  }

  &--size-medium {
    @include border-radius;
    @include padding-button-small;
  }
  &--size-full {
    width: 100%;
    text-align: center;
  }

  &--size-round {
    @include title-small;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    user-select: none;
    border-radius: $border-radius-circle;
    padding-right: 0;
    padding-left: 0;
    width: $space-xxl;
    height: $space-xxl;
    text-align: center;

    .button__icon {
      margin: 0;
    }
  }
  &--size-square {
    @include text-body;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    text-align: center;
    user-select: none;
    border-radius: $border-radius-m;
    padding-right: 0;
    padding-left: 0;
    width: $space-xl;
    height: $space-xl;
    padding: $space-xxs;
    text-align: center;
  }

  @mixin type-text {
    @include text-small;
    background-color: transparent;
    text-transform: uppercase;
    color: inherit;
    padding: 0;
    margin: 0;
    border: none;
  }

  &--type-text {
    @include type-text;
    font-weight: $font-weight-normal;
  }

  &--type-text-active {
    @include type-text;
    font-weight: $font-weight-black;
  }

  &--type-primary {
    background-color: $color-primary;
    border-color: $color-primary;
    color: $color-primary-inverse;

    &:hover {
      background-color: $color-primary-active;
      border-color: $color-primary-active;
    }
  }

  &--type-secondary {
    background-color: $color-secondary;
    border-color: $color-secondary;
    color: $color-secondary-inverse;

    &:hover {
      // todo : garder la border peut-être? ou rendre l'effet un peu plus légeer
      background-color: $color-secondary-active;
      border-color: $color-secondary-active;
    }

    &:focus {
      @include focus($color-secondary);
    }
  }

  &--type-tertiary {
    background-color: $color-tertiary;
    border-color: $color-tertiary;
    color: $color-tertiary-inverse;

    &:hover {
      // todo : garder la border peut-être? ou rendre l'effet un peu plus légeer
      background-color: $color-tertiary-active;
      border-color: $color-tertiary-active;
    }

    &:focus {
      @include focus($color-tertiary);
    }
  }

  &--type-positive {
    background-color: $color-success;
    border-color: $color-success;
    color: $color-success-inverse;

    &:hover {
      background-color: $color-success-active;
      border-color: $color-success-active;
    }

    &:focus {
      @include focus($color-success);
    }
  }

  &--type-warning {
    background-color: $color-warning;
    border-color: $color-warning;
    color: $color-warning-inverse;

    &:hover {
      background-color: $color-warning-active;
      border-color: $color-warning-active;
    }

    &:focus {
      @include focus($color-warning-active);
    }
  }

  &--type-negative {
    background-color: $color-danger;
    border-color: $color-danger;
    color: $color-danger-inverse;

    &:hover {
      background-color: $color-danger-active;
      border-color: $color-danger-active;
    }

    &:focus {
      @include focus($color-danger-active);
    }
  }

  &--type-neutral {
    color: $color-neutral-20;
    background-color: $background-color-base;
    border-color: $background-color-softest;

    &:hover {
      background-color: $background-color-softest;
    }

    &:focus {
      @include focus($background-color-softest);
    }
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  &--loading {
    opacity: 0.7;
  }
}

.button__icon {
  &--left {
    margin-right: $space-xs;
  }
  &--right {
    margin-left: $space-xs;
  }
}

button {
  outline: none !important;
}
</style>
<docs>
## Colors
  ```jsx
    <mks-button tag="button">Button default</mks-button>
    <mks-button tag="button" type="secondary">Button secondary</mks-button>
    <mks-button tag="button" type="tertiary">Button tertiary</mks-button>
    <mks-button tag="button" type="neutral">Button neutral</mks-button>
    <br>
    <mks-button type="positive">Button positive</mks-button>
    <mks-button type="warning">Button warning</mks-button>
    <mks-button type="negative">Button negative</mks-button>
  ```

## Sizes
```jsx
    <mks-button size="round" icon="menu"></mks-button>
    <mks-button size="small">Button small</mks-button>
    <mks-button size="medium">Button medium</mks-button>
    <mks-button size="default">Button default</mks-button>
    <mks-button size="full">Button full</mks-button>
  ```

## With icons
```jsx
    <mks-button size="round" icon="bell"></mks-button>
    <br>
    <mks-button icon="menu" size="small">Menu</mks-button>
    <mks-button icon="menu" icon-right size="small">Menu</mks-button>
    <br>
    <mks-button icon="star">Sauvegarder</mks-button>
    <mks-button icon="star" icon-right>Sauvegarder</mks-button>
  ```

  ## States
```jsx
    <mks-button v-bind:disabled="true">Disabled</mks-button>
    <br>
    <mks-button v-bind:loading="true">
        Load 
        <template v-slot:loading>Loading...</template>
    </mks-button>
    <br>
    <mks-button v-bind:loading="true">
        Load 
    </mks-button>
    <br>
    <mks-button v-bind:loading="true" size="small">
        Load 
    </mks-button>
  ```
</docs>
