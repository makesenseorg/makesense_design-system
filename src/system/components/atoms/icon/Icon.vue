<template>
  <svg
    :class="`icon icon-${getIcon} icon--color-${color}`"
    role="img"
    xmlns="http://www.w3.org/2000/svg"
    :width="size"
    :height="size"
    :viewBox="`0 0 ${defaultSize} ${defaultSize}`"
    fill="none"
    :stroke-width="strokeWidth"
    stroke-linecap="round"
    stroke-linejoin="round"
    v-html="path"
  ></svg>
</template>

<script>
import feather from "feather-icons";
/**
 * Nice icons to make it more fun ✨ (based on feather)
 * @version 0.1.0
 */
export default {
  name: "MksIcon",
  props: {
    /**
     * Type of icon
     */
    type: {
      type: String,
      required: true
    },
    /**
     * Description of the icon for screen readers
     */
    alt: {
      type: String,
      required: false
    },
    /**
     * Color of the icon amongs theme colors, inherits from parent by default.
     */
    color: {
      type: String,
      default: "inherit",
      validator: function(value) {
        return (
          [
            "inherit",
            "primary",
            "secondary",
            "tertiary",
            "neutral",
            "positive",
            "warning",
            "negative",
            "currentColor"
          ].indexOf(value) !== -1
        );
      }
    },
    size: {
      type: [String, Number],
      default: "24"
    },
    strokeWidth: {
      type: String,
      default: "2"
    }
  },
  data() {
    return {
      defaultSize: 24
    };
  },
  computed: {
    getIcon: function() {
      if (this.type === "close") return "x";
      if (this.type === "arrowUp") return "arrow-up";
      if (this.type === "arrowDown") return "arrow-down";
      if (this.type === "arrowLeft") return "arrow-left";
      if (this.type === "arrowRight") return "arrow-right";
      return this.type;
    },
    path() {
      return `<title id="${this.type}">${this.alt || ""}</title>${
        feather.icons[this.type]
      }`;
    }
  }
};
</script>

<style lang="scss" scoped>
.icon {
  stroke: currentColor;
  vertical-align: middle;

  &--color-neutral {
    stroke: $color-neutral-70;
  }

  &--color-primary {
    stroke: $color-primary;
  }

  &--color-secondary {
    stroke: $color-secondary;
  }

  &--color-tertiary {
    stroke: $color-tertiary;
  }

  &--color-positive {
    stroke: $color-success;
  }

  &--color-warning {
    stroke: $color-warning;
  }

  &--color-negative {
    stroke: $color-danger;
  }
}
</style>
<docs>
## Icons
[List of 286 available icons](https://feathericons.com/)

Some examples :

```jsx
<p>
  <mks-icon type="activity" alt="Activités"></mks-icon> 
  <mks-icon type="alert-triangle" alt="Attention, il fait froid!"></mks-icon> 
  <mks-icon type="arrow-left" alt="A gauche"></mks-icon>
  <mks-icon type="arrow-left-circle" alt="Retourner"></mks-icon>
  <mks-icon type="bell" alt="Notifications"></mks-icon> 
  <mks-icon type="check"></mks-icon> 
  <mks-icon type="clock"></mks-icon> 
  <mks-icon type="eye-off"></mks-icon> 
  <mks-icon type="filter"></mks-icon> 
  <mks-icon type="heart"></mks-icon> 
  <mks-icon type="list"></mks-icon> 
  <mks-icon type="lock"></mks-icon>
  <mks-icon type="map-pin"></mks-icon>
  <mks-icon type="menu"></mks-icon>
  <mks-icon type="save"></mks-icon>
  <mks-icon type="star"></mks-icon> 
  <mks-icon type="thumbs-up"></mks-icon> 
  <mks-icon type="user"></mks-icon>
  <mks-icon type="zoom-in"></mks-icon>
</p>

<p>
  <mks-icon type="star" alt="petite étoile" v-bind:size="12" color="primary"></mks-icon> 
  <mks-icon type="star" alt="étoile couleur secondaire" size="50" color="secondary"></mks-icon> 
</p>
```
</docs>
