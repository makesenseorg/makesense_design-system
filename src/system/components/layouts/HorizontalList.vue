<template>
  <component :is="tag" class="horizontal-list">
    <slot></slot>
  </component>
</template>
<script>
/**
 * Displays items in a horizontal line
 * @version 0.2.0
 */
export default {
  name: "MksHorizontalList",
  release: "0.2.0",
  props: {
    tag: {
      type: String,
      default: "div"
    }
  }
};
</script>
<style lang="scss" scoped>
.horizontal-list {
  @include flex;
  align-items: center;

  & > * {
    list-style-type: none;
    margin-right: $space-base;

    &:last-child {
      margin-right: 0;
    }
  }
}
</style>
<docs>

```jsx
<mks-horizontal-list>
  <mks-tag>Tag 1</mks-tag>
  <mks-tag>Tag 2</mks-tag>
  <mks-tag>Tag 3</mks-tag>
</mks-horizontal-list>

```


</docs>
