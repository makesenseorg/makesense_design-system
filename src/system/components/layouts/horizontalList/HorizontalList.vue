<template>
  <component
    :is="tag"
    :class="`horizontal-list horizontal-list--spacing-${spacing}`"
  >
    <slot></slot>
  </component>
</template>
<script>
/**
 * Deprecated: Use Spacer instead - Displays items in a horizontal line
 * @version 0.2.0
 * @deprecated
 */
export default {
  name: "MksHorizontalList",
  status: "deprecated",
  release: "0.2.0",
  props: {
    tag: {
      type: String,
      default: "div"
    },
    spacing: {
      type: String,
      default: "default",
      validator: function(value) {
        return ["default", "none", "small", "large"].indexOf(value) !== -1;
      }
    }
  }
};
</script>
<style lang="scss" scoped>
.horizontal-list {
  @include flex;
  align-items: center;

  &--spacing-small > * {
    margin-right: $space-xs;
  }
  &--spacing-default > * {
    margin-right: $space-m;
  }
  &--spacing-large > * {
    margin-right: $space-l;
  }

  & > * {
    list-style-type: none;

    &:last-child {
      margin-right: 0;
    }
  }
}
</style>
<docs>

Deprecated: Use <mks-link to="/layouts/mksspacer">Spacer</mks-link> instead

```jsx
<mks-horizontal-list spacing="none">
  <mks-tag>Tag 1</mks-tag>
  <mks-tag>Tag 2</mks-tag>
  <mks-tag>Tag 3</mks-tag>
</mks-horizontal-list>

<mks-horizontal-list spacing="small">
  <mks-tag>Tag 1</mks-tag>
  <mks-tag>Tag 2</mks-tag>
  <mks-tag>Tag 3</mks-tag>
</mks-horizontal-list>

<mks-horizontal-list>
  <mks-tag>Tag 1</mks-tag>
  <mks-tag>Tag 2</mks-tag>
  <mks-tag>Tag 3</mks-tag>
</mks-horizontal-list>

<mks-horizontal-list spacing="large">
  <mks-tag>Tag 1</mks-tag>
  <mks-tag>Tag 2</mks-tag>
  <mks-tag>Tag 3</mks-tag>
</mks-horizontal-list>

```


</docs>
