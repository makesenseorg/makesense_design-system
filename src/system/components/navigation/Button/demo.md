## Button types

Use a primary button to draw the users attention to important actions. Use default buttons for less important actions.

A danger button should be used only for destructive actions.

```
<ds-button>Default</ds-button>
<ds-button primary>Primary</ds-button>
<ds-button secondary>Secondary</ds-button>
<ds-button danger>Danger</ds-button>
```

## Ghost buttons

Use a ghost button for secondary actions.

```
<ds-button ghost>Default</ds-button>
<ds-button ghost primary>Primary</ds-button>
<ds-button ghost secondary>Secondary</ds-button>
<ds-button ghost danger>Danger</ds-button>
```

## Button sizes

Use different sizes to create hierarchy.

```
<ds-button size="small">Small</ds-button>
<ds-button>Base</ds-button>
<ds-button size="large">Large</ds-button>
```

## Fullwidth button

Use a fullwidth button for important actions.

```
<ds-button primary fullwidth>Fullwidth button</ds-button>
```

## Button states

A button can take different states.

```
<ds-button>Default state</ds-button>
<ds-button disabled>Disabled state</ds-button>
<ds-button hover>Hover state</ds-button>
```

## Icon buttons

Add an icon to a button to help the user identify the button's action.

```
<ds-button icon="plus" primary>Click me</ds-button>
<ds-button icon-right="plus">Click me</ds-button>
<ds-button icon="plus" primary></ds-button>
<ds-button icon="plus" ghost></ds-button>
```

## Button as links

Provide a path to the button. You can pass a url string or a Vue router path object.

```
<ds-button path="/navigation">Click me</ds-button>
<ds-button :path="{ name: 'Navigation' }">Click me</ds-button>
```