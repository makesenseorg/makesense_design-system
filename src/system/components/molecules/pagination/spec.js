import { shallowMount, createLocalVue } from "@vue/test-utils";
import Comp from "./Pagination.vue";
import Button from "../../atoms/button/Button.vue";
import Icon from "../../atoms/icon/Icon.vue";
import HorizontalList from "../../layouts/horizontalList/HorizontalList.vue";

const localVue = createLocalVue();
localVue.component("MksButton", Button);
localVue.component("MksIcon", Icon);
localVue.component("MksHorizontalList", HorizontalList);

describe("Pagination.vue", () => {
  it("matches snapshot", () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 10,
        value: 5
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("pads 2 pages by default", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 50
      }
    });
    expect(wrapper.text().replace(/\s+/g, ",")).toBe(
      "1,2,3,...,48,49,50,51,52,...,98,99,100"
    );
  });

  it("pads pages according to the pad value", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 50,
        pad: 1
      }
    });
    expect(wrapper.text().replace(/\s+/g, ",")).toBe(
      "1,2,...,49,50,51,...,99,100"
    );
  });

  it("pads the first/last pages according to the padEnd value", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 50,
        padEnd: 1
      }
    });
    expect(wrapper.text().replace(/\s+/g, ",")).toBe(
      "1,2,...,48,49,50,51,52,...,99,100"
    );
  });

  it("pads the current page according to the padAround value", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 50,
        padAround: 1
      }
    });
    expect(wrapper.text().replace(/\s+/g, ",")).toBe(
      "1,2,3,...,49,50,51,...,98,99,100"
    );
  });

  it("doesn't show an ellipsis for a single page", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 5,
        pad: 1
      }
    });
    expect(wrapper.text().replace(/\s+/g, ",")).toBe("1,2,3,4,5,6,...,99,100");
  });

  it("changes the current page on click", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 5,
        pad: 1
      }
    });
    expect(wrapper.text().replace(/\s+/g, ",")).toBe("1,2,3,4,5,6,...,99,100");
  });

  it("goes to the previous page when clicked", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 5,
        pad: 1
      }
    });
    const button = wrapper.find(".pagination__button--prev");
    button.vm.$emit("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().change[0]).toEqual([4]);
  });

  it("goes to the next page when clicked", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 5,
        pad: 1
      }
    });
    const button = wrapper.find(".pagination__button--next");
    button.vm.$emit("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().change[0]).toEqual([6]);
  });

  it("goes to the selected page when clicked", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 5,
        pad: 1
      }
    });
    const button = wrapper.find(".pagination__item");
    button.vm.$emit("click");
    await wrapper.vm.$nextTick();
    expect(wrapper.emitted().change[0]).toEqual([1]);
  });

  it("disables the 'previous' button on the first page", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 1,
        pad: 1
      }
    });
    const button = wrapper.find(".pagination__button--prev");
    expect(button.classes()).toContain("disabled");
  });

  it("disables the 'next' button on the last page", async () => {
    const wrapper = shallowMount(Comp, {
      localVue,
      propsData: {
        total: 100,
        value: 100,
        pad: 1
      }
    });
    const button = wrapper.find(".pagination__button--next");
    expect(button.classes()).toContain("disabled");
  });
});
