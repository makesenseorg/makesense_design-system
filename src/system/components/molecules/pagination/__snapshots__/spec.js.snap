// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.vue matches snapshot 1`] = `
<mks-horizontal-list-stub
  class="pagination"
  spacing="small"
  tag="div"
>
  <mks-button-stub
    class="pagination__button pagination__button--prev"
    inputtype=""
    size="square"
    tag="div"
    type="neutral"
  >
    <mks-icon-stub
      color="inherit"
      library="feather"
      size="24"
      strokewidth="2"
      type="arrow-left"
    />
  </mks-button-stub>
   
  <mks-horizontal-list-stub
    class="pagination__list"
    spacing="small"
    tag="ul"
  >
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        1
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        2
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        3
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        4
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item pagination__item--current"
      inputtype=""
      size="square"
      tag="div"
      type="secondary"
    >
      <span>
        
        5
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        6
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        7
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        8
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        9
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
    <mks-button-stub
      class="pagination__item"
      inputtype=""
      size="square"
      tag="div"
      type="neutral"
    >
      <span>
        
        10
      
      </span>
       
      <div
        class="effect"
      />
    </mks-button-stub>
  </mks-horizontal-list-stub>
   
  <mks-button-stub
    class="pagination__button pagination__button--next"
    inputtype=""
    size="square"
    tag="div"
    type="neutral"
  >
    <mks-icon-stub
      color="inherit"
      library="feather"
      size="24"
      strokewidth="2"
      type="arrow-right"
    />
  </mks-button-stub>
</mks-horizontal-list-stub>
`;
